### PR TITLE
Add OIDC prefix support for impersonation

### PR DESCRIPTION
**Pull request by sarataha**
Closes https://github.com/weaveworks/weave-gitops/issues/3585

**What changed?**
This PR adds support for configuring OIDC username and groups prefixes. Changes:
* Add two new flags, `oidc-username-prefix` and `oidc-groups-prefix`.
* Read prefix values out of OIDCConfig secret.
* Update the impersonating client to add the prefixes.

**Why was this change made?**
To allow for impersonation with OIDC prefixes.

**How did you validate the change?**
Added a test case for `ConfigWithPrincipal` to ensure prefixes were added correctly.

### CODE REVIEW

File: pkg/kube/config_getter.go [hunk 1]
```diff
@@ -7,6 +7,11 @@ import (
 	"k8s.io/client-go/rest"
 )
 
+type OIDCPrefixes struct {
```
Comment (bigkevmcd) [inline pkg/kube/config_getter.go:10]: This is exported, so should have a doc-comment.

But, it also feels like these are not `OIDCPrefixes` at this point, yes, we might use these to match against groups that come from OIDC, but these are not related to OIDC?

Is there a better name?

File: pkg/kube/config_getter.go [hunk 2]
```diff
@@ -20,14 +25,15 @@ var _ ConfigGetter = &ImpersonatingConfigGetter{}
 // principal and if it finds one, it configures the *rest.Config to impersonate
 // that principal. Otherwise it returns a copy of the base config.
 type ImpersonatingConfigGetter struct {
-	insecure bool
-	cfg      *rest.Config
+	insecure     bool
+	cfg          *rest.Config
+	oIDCPrefixes OIDCPrefixes
```
Comment (bigkevmcd) [inline pkg/kube/config_getter.go:30]: Should this be `OIDCPrefixes` ?

File: pkg/kube/config_getter_test.go [hunk 3]
```diff
@@ -78,3 +78,38 @@ func TestImpersonatingConfigGetterNoPrincipalInContext(t *testing.T) {
...
+	if cfg.Impersonate.UserName != expectedUser {
+		t.Fatalf("cfg username didn't match expected: %s", cfg.Impersonate.UserName)
+	}
+
+	expectedGroups := []string{"prefix-group1", "prefix-group2"}
+	if diff := cmp.Diff(expectedGroups, cfg.Impersonate.Groups); diff != "" {
+		t.Fatalf("cfg groups didn't match expected:\n%s", diff)
+	}
+
+	// Test with multiple calls.
```
Comment (bigkevmcd) [inline pkg/kube/config_getter_test.go:108]: You don't need to do this 10 times, you can just do it again :-)

**Review by bigkevmcd — CHANGES_REQUESTED** [review]

File: pkg/kube/config_getter.go (hunk 1 above)
Comment (sarataha) [reply pkg/kube/config_getter.go:10]: Makes sense to me 🙂 Is `UserPrefixes` suitable?

File: pkg/kube/config_getter_test.go (hunk 3 above)
Comment (sarataha) [reply pkg/kube/config_getter_test.go:108]: Yes thank you for point this out 🙂 🙏

**Review by bigkevmcd — APPROVED** [review]
Looks good to me.

This will require changes to the enterprise product to pass these through too.

File: cmd/gitops-server/cmd/cmd.go [hunk 4]
```diff
@@ -187,7 +191,12 @@ func runCmd(cmd *cobra.Command, args []string) error {
 
 	ctx := context.Background()
 
-	cl, err := cluster.NewSingleCluster(cluster.DefaultCluster, rest, scheme, cluster.DefaultKubeConfigOptions...)
+	oidcPrefixes := kube.UserPrefixes{
+		UsernamePrefix: options.OIDC.UsernamePrefix,
+		GroupsPrefix:   options.OIDC.GroupsPrefix,
+	}
```
Comment (foot) [inline cmd/gitops-server/cmd/cmd.go:197]: This should also incorportate authServer.AuthConfig.OIDCConfig.GroupsPrefix/UsernamePrefix I think?

Given https://github.com/weaveworks/weave-gitops/blob/cb8585c028ca6287220c49e9c3643d60f64a645d/pkg/server/auth/init.go#L40-L41

We should merge them w/ the secret taking higher precedence and show a similar logging message.

Comment (sarataha) [reply cmd/gitops-server/cmd/cmd.go:197]: Got it 👍 Thank you 🙂

File: cmd/gitops-server/cmd/cmd.go [hunk 5]
```diff
@@ -187,7 +191,22 @@ func runCmd(cmd *cobra.Command, args []string) error {
...
+
+	// Incorporate values from authServer.AuthConfig.OIDCConfig
+	if authServer.AuthConfig.OIDCConfig.UsernamePrefix != "" {
+		log.V(logger.LogLevelWarn).Info("OIDC client configured by both CLI and secret. Secret values will take precedence.")
```
Comment (foot) [inline cmd/gitops-server/cmd/cmd.go:201]:
```suggestion
		log.V(logger.LogLevelWarn).Info("OIDC username prefix configured by both CLI and secret. Secret values will take precedence.")
```

**Review by foot — APPROVED** [review]
